### DISCUSSION TITLE
Consider forces direction difference between VASP & LAMMPS results?

### DISCUSSION THREAD

**Discussion by TSMtang** [mir-group/flare#380] [Q&A] [answered]
Dear Yu Xie,

I've been using flare otf training with sgp. I noticed one problem during this process.

My structure was relaxed with VASP before being used for otf training, I used VASP setup ISIF=3 which means the unit cell length and angles are also relaxed. The final unit cell is described in 9 parameters. The same structure in LAMMPS will describe the unit cell in 6 parameters, which means there is a coordinate transformation between the VASP & LAMMPS structure. 

I checked your source code about how you defined the "flare_atom" object and how you used the "get_forces" method from ASE to extract the forces information from both VASP & LAMMPS results. 

My question: Is the difference in the coordinate system considered? It seems to me that this coordinate difference will cause forces direction difference because it's rotated. When otf training is trying to minimize the "forces mae", does it consider this rotation?

Best regards
Shengming Tang

**Comment by YuuuXie** [comment]
The stress tensor is 3x3 but only has 6 degrees of freedom xx, yy, zz, xy, xz, yz. For example, for a given ase atoms you can choose `get_stresses(voigt=True)` or `get_stresses(voigt=False)` which it will return 3x3 matrix or a vector of length 6 correspondingly.

There might be a wrapping by lammps when you input your vasp structure. But it does not matter since it is not a global "rotation" but just shifting the periodic boundary, so the forces on each atom won't change.

Also the otf training is not trying to minimize the force MAE, the active learning is not training a neural network and minimizing the loss function. It just tries to detect and collect uncertain configurations

**Reply by TSMtang** [reply]
Dear Yu Xie,

First of all, thank you for your explanation and your patience, and thank you for your quick reply.

1. You are totally correct that for stress tensor, the 3X3 matrix has only 6 degrees of freedom. That's because the matrix is symmetric: Sxy = Syx and so on. But when comes to the lattice parameters, it will be different. 

2. In vasp, 9 parameters are used to describe the unit cell and it's truly 9 degree of freedoms if ISIF=3 is used: [ISIF](https://www.vasp.at/wiki/index.php/ISIF). On the other hand, we all know that 6 parameters are enough to describe a unit cell (a, b, c, alpha, gamma, beta) and this 6 parameters strategy is commonly used for example in .cif file and in lammps. 

The wrap of lammps will take the vasp input structure, convert the unit cell into 6 parameters unit cell. During this step, 3 degrees of freedom are lost which I assume to be the rotation of the unit cell on three dimensions. You are right that this is purely a shifting of the periodic boundary in vasp, but lammps will create the new cartesian coordinate system which has x-axis on the "a" vector of the unit cell. This means in vasp and lammps, the two cartesian coordinate systems are different, of course, all the atom positions will be different and I assume all forces directions will be different.

I understand that you didn't face this problem because, in your Pt-H2 paper, you considered the surface material and didn't need to relax the unit cell. But this rotation can be verified by simply reading the vasp structure and writing the lammps structure by ASE. (I can provide you example vasp structure if you want).

I noticed this problem because I already trained a sgp model and map the force field for lammps. I'm testing the accuracy of the force field on other systems (some test data). I just want to make sure I'm calculating the force error correctly.

3. Thank you for mentioning it's not minimizing the force MAE, I get it that otf is trying to collect DFT data and select local environments as training data for the sgp model, then the sgp regression will make sure all the training data is in some uncertainty range. Please correct me if I'm wrong.

Thank you for your answer
Shengming Tang

**Reply by YuuuXie** [reply, marked as answer]
Thank you for the explanation. We don't do a global rotation within the flare code, so if you have two cells where one is rotated and the other is not, then you will get mismatched forces. I think for now the best way is to alway preprocessing your vasp relaxed structure to align the x-axis, before you start training and also the testing.
